File: 2-D Dynamic Programming/3573. Best Time to Buy and Sell Stock V/solution.py
class Solution:
    def maximumProfit(self, prices: List[int], K: int) -> int:
        n = len(prices)

        # tx_type=0 no tx
        # tx_type=1 sell
        # tx_type=2 short sell
        memo = [[[None]*3 for _ in range(K+1)] for _ in range(n)]

        def dfs(i, k, tx_type):
            if i >= n:
                return 0 if k >= 0 and tx_type == 0 else -inf
            if memo[i][k][tx_type] is None:
                memo[i][k][tx_type] = dfs(i+1, k, tx_type)
                if k > 0:
                    if tx_type == 0:
                        memo[i][k][tx_type] = max(memo[i][k][tx_type], dfs(i+1, k, 1) - prices[i])
                        memo[i][k][tx_type] = max(memo[i][k][tx_type], dfs(i+1, k, 2) + prices[i])
                    else:
                        if tx_type == 1:
                            memo[i][k][tx_type] = max(memo[i][k][tx_type], dfs(i+1, k-1, 0) + prices[i])
                        if tx_type == 2:
                            memo[i][k][tx_type] = max(memo[i][k][tx_type], dfs(i+1, k-1, 0) - prices[i])
            return memo[i][k][tx_type]

        return dfs(0, K, 0)
    
# iterative method
class Solution:
    def maximumProfit(self, prices: List[int], k: int) -> int:
        bought = [-inf] * k
        res = [0] * (k + 1)
        sold = [0] * k
        for p in prices:
            for kk in range(k, 0, -1):
                res[kk] = max(res[kk], bought[kk - 1] + p, sold[kk - 1] - p)
                bought[kk - 1] = max(bought[kk - 1], res[kk - 1] - p)
                sold[kk - 1] = max(sold[kk - 1], res[kk - 1] + p)
        return max(res)

File: 2-D Dynamic Programming/3573. Best Time to Buy and Sell Stock V/solution.cpp
#include <vector>
#include <numeric>

using namespace std;

class Solution {
public:
    long long maximumProfit(vector<int>& prices, int k) {
        int n = prices.size();

        vector<vector<vector<long long>>>memo(n+1, vector<vector<long long>>(k+1, vector<long long>(3, -1)));

        function<long long (int i, int k, int type)> dfs = [&](int i, int k, int type) -> long long {
            if (i >= n) return (k >= 0 && type == 0 ? 0 : LLONG_MIN/2);

            if (memo[i][k][type] == -1) {
                memo[i][k][type] = dfs(i+1, k, type);
                if (k > 0) {
                    if (type == 0) {
                        long long res1 = dfs(i+1, k, 1) - prices[i];
                        long long res2 = dfs(i+1, k, 2) + prices[i];
                        memo[i][k][type] = max(memo[i][k][type], max(res1, res2));
                    } else if (type == 1) /* normal tx */ {
                        memo[i][k][type] = max(memo[i][k][type], dfs(i+1, k-1, 0) + prices[i]);
                    } else if (type == 2) /* sell short */ {
                        memo[i][k][type] = max(memo[i][k][type], dfs(i+1, k-1, 0) - prices[i]);
                    }
                }
            }

            return memo[i][k][type];
        };

        return dfs(0, k, 0);
    }
};


class Solution {
public:
    long long maximumProfit(vector<int>& prices, int k) {
        static const long long NEG_INF = LLONG_MIN/2;

        vector<long long> bought(k, NEG_INF);
        vector<long long> sold(k, NEG_INF);
        vector<long long> result(k + 1, NEG_INF);
        result[0] = 0;
        for (const auto& x : prices) {
            for (int kk = k - 1; kk >= 0; --kk) {
                if (bought[kk] != NEG_INF) {
                    result[kk + 1] = max(result[kk + 1], bought[kk] + x);
                }
                if (sold[kk] != NEG_INF) {
                    result[kk + 1] = max(result[kk + 1], sold[kk] - x);
                }
                if (result[kk] != NEG_INF) {
                    bought[kk] = max(bought[kk], result[kk] - x);
                    sold[kk] = max(sold[kk], result[kk] + x);
                }
            }
        }

        long long res = LLONG_MIN;
        for (const auto rslt : result) {
            res = max(res, rslt);
        }
        return res;
    }
};